### PR TITLE
Skip articles in KeywordTitle assessment

### LINKED ISSUE #1003: Keyword is not found if preceded by Spanish question mark
## What did you expect to happen?
Keywords to be found regardless of they are preceded by a Spanish question mark or not.

## What happened instead?
Spanish questions start with an upside-down question mark attached to the first word of the sentence. Because of this question mark, the keyword is not recognized if it is at the beginning of a question.  

schermafbeelding 2017-01-02 om 14 11 26

## How can we reproduce this behavior?
Start a sentence with a Spanish question mark, directly followed by your focus keyword. 

## Technical info

* Platform: WordPress
* Platform version: 4.7
* Yoast SEO version: 4.0.2

### LINKED ISSUE #1313: Improve the keyword in title assessment
### Description
We should filter stopwords so that keywords preceded by e.g. `the` are still considered to be at the "beginning" of the title.

### PR DESCRIPTION
**Pull request by nataliashitova**
## Summary

This PR can be summarized in the following changelog entry:

* Skips articles in the beginning of the SEO title when checking if the keyword is present in (the beginning of) the title.
* Checks for the keyword in the title only after a title has been set.

## Relevant technical choices:

* The assessment now skips all articles specified for a given locale when assessing if a keyword is present in the beginning of the title. This means that if a user has a title "A dog" and wants to rank for "dog" the assessment will return a green "good" score.
* This also means that if a user has a title "A the dog" s/he will receive a high score.
* An additional requirement was added to the paper to receive this assessment: The paper has to have a title. It was done to prevent situations, when a user receives a feedback "Please create an SEO title." from TitleWidth assessment and then "The focus keyword '%1$s' does not appear in the SEO title." from KeywordTitle assessment.
* Because of that additional requirement, I had to edit the specs.
* Added an inverted question mark to the word boundary processing and trimming punctuation to account for cases like title: "¿Cómo hacer guacamole?" ranking for "cómo". As a side effect this also fixed the bug described in #1003.

## Test instructions

This PR can be tested by following these steps:

### Browserified

- check out the branch
- use the browserfied example. Don't forget to run a grunt build:js
- set locale to any language or leave default
- add a keyword and a title that starts directly with that keyword - the assessment should return a green bullet "The SEO title contains the focus keyword, at the beginning which is considered to improve rankings."
- add an article in the beginning of your title (e.g., "a", "an", "the" for English) - the assessment result should remain the same
- add a different non-article word in the beginning of your title - the assessment should return an orange bullet "The SEO title contains the focus keyword, but it does not appear at the beginning; try and move it to the beginning."
- remove the keyword from the title - the assessment should return a red bullet "The focus keyword '%1$s' does not appear in the SEO title."

### Wordpress

- Check out `stories/9630-add-titleKeywordAssessment-to-SEOAssessor` in wordpress-seo.
- Run yarn install in the plugin's directory.
- Run yarn add "git+https://github.com/Yoast/yoastseo.js#stories/1313-improve-keyword-title-assessment" in the plugin's directory.
- Navigate to node_modules/yoastseo and run yarn install and grunt build:js.
- Navigate back to the plugin's directory and run grunt build:js.
- Test the same functionality as described in the test instructions for Browserified.

Alternatively, you can use the beta script to test the same combination of branches.

Fixes #1313, and fixes #1003

### CODE REVIEW

File: src/assessments/seo/titleKeywordAssessment.js [hunk 1]
```diff
@@ -1,46 +1,113 @@
...
-	if ( keywordMatches.matches > 0 && keywordMatches.position > 0 ) {
-		score = 6;
-		text = i18n.dgettext( "js-text-analysis", "The SEO title contains the focus keyword, but it does not appear at the beginning;" +
-			" try and move it to the beginning." );
+	/**
+	 * Calculates the result based on the keyphraseLength research.
+	 *
+	 * @returns {Object} object with score and text.
+	 */
+	calculateScore() {
```
Comment (manuelaugustin) [inline src/assessments/seo/titleKeywordAssessment.js]: calculateResult? (in line with the previous assessments that have been altered like this)

File: src/assessments/seo/titleKeywordAssessment.js [hunk 2]
```diff
@@ -1,46 +1,113 @@
...
+	 * @param {string} keyword The keyword of the paper
+	 * @param {Object} i18n The i18n-object used for parsing translations.
+	 *
+	 * @returns {string} Text feedback.
+	 */
+	translateScore( resultText, requiresKeyword, keyword, i18n ) {
+		if ( requiresKeyword ) {
+			return i18n.sprintf( i18n.dgettext(
+				"js-text-analysis",
+				/* Translators: %1$d expands to the keyphrase. */
```
Comment (manuelaugustin) [inline src/assessments/seo/titleKeywordAssessment.js]: I'm not entirely sure how these translator comments are parsed, but I suspect that now they'll also get this comment in the cases when this variable isn't present (good & okay). Perhaps we should check whether that's an issue.

File: src/researches/italian/functionWords.js [hunk 3]
```diff
@@ -245,6 +245,9 @@ module.exports = function() {
 			indefinitePronouns, correlativeConjunctions, subordinatingConjunctions, interrogativeDeterminers, interrogativePronouns,
 			interrogativeAdverbs, locativeAdverbs, miscellaneous, pronominalAdverbs, recipeWords, timeWords, vagueNouns ),
 
+		// Articles are used in the keywordTitle assessment
```
Comment (manuelaugustin) [inline src/researches/italian/functionWords.js]: End sentence with period.

File: src/researches/portuguese/functionWords.js [hunk 4]
```diff
@@ -167,6 +167,9 @@ module.exports = function() {
 			interrogativeProAdverbs, locativeAdverbs, otherAuxiliaries, copula, subordinatingConjunctions, interviewVerbs, additionalTransitionWords,
 			delexicalizedVerbs, interjections, recipeWords, timeWords, vagueNouns, miscellaneous, titles ),
 
+		// Articles are used in the keywordTitle assessment
```
Comment (manuelaugustin) [inline src/researches/portuguese/functionWords.js]: End sentence with period.

File: src/researches/spanish/functionWords.js [hunk 5]
```diff
@@ -290,6 +290,9 @@ module.exports = function() {
 		// These word categories cannot intervene between an auxiliary and a corresponding passive participle.
 		cannotBeBetweenPassiveAuxiliaryAndParticiple: [].concat( copulaEstar, copulaEstarInfinitive ),
 
+		// Articles are used in the keywordTitle assessment
```
Comment (manuelaugustin) [inline src/researches/spanish/functionWords.js]: End sentence with period.

**Comment by manuelaugustin** [comment]
CR 🚧

* Some minor documentation style issues, see comments.

* Code coverage for findKeywordInPageTitle.js has decreased

* I was thinking again about the changelog description. I think it would be good to mention that the titleKeyword assessment is now only shown when there's a title in the changelog, since it's a user-facing change. So something like "The keyword in the title is only checked after a title has been set."

* Wasn't there some issue about the Spanish ¿ that is fixed now? Because then you can also add that to the fixed issues.

File: src/assessments/seo/titleKeywordAssessment.js (hunk 1 above)
Comment (nataliashitova) [reply src/assessments/seo/titleKeywordAssessment.js]: Agreed.

File: src/assessments/seo/titleKeywordAssessment.js [hunk 6]
```diff
@@ -1,46 +1,113 @@
...
+				resultText
+			), keyword );
+		}
 
-	return assessmentResult;
-};
+		return i18n.dgettext( "js-text-analysis", resultText );
```
[the remark this answers is not in the document]
Comment (nataliashitova) [reply src/assessments/seo/titleKeywordAssessment.js]: Does not seem to matter (and it used to be without a sprinf wrapper in the previous stable version of the assessment), but maybe indeed a safer way.

**Comment by nataliashitova** [comment]
@manuelaugustin 

> Some minor documentation style issues, see comments.
Fixed

> Code coverage for findKeywordInPageTitle.js has decreased
I am not seeing it. I checked build print-out and coverage/lcov-report/index.html, both the previous version and the current version give 100% coverage.

> I was thinking again about the changelog description. I think it would be good to mention that the titleKeyword assessment is now only shown when there's a title in the changelog, since it's a user-facing change. So something like "The keyword in the title is only checked after a title has been set."
Fixed

> Wasn't there some issue about the Spanish ¿ that is fixed now? Because then you can also add that to the fixed issues.
I was citing that issue in the description of the pull request. Now I also put it in `Fixes`

> There seem to be some merge conflicts which need to be resolved.
Resolved

**Comment by manuelaugustin** [comment]
CR 🚧

* See the screenshot of the code coverage of `findKeywordInPageTitle.js` as I see it in the report (the implicit else condition isn't covered).

* Github seems to be confused again: recent changes in develop (Flesch reading ease, PR about the keywords consisting only of special characters) are also in this PR. If all the branches have been pulled/merged/pushed correctly, probably  just changing the base branch to develop and changing it back again to feature/recalibration should fix the issue.

* There are some merge conflicts.

* I changed the `fixes` in the issue description; you need to add `fixes` before every issue number for Github to recognize it (see https://help.github.com/articles/closing-issues-using-keywords/).

screen shot 2018-05-07 at 11 26 32

**Comment by nataliashitova** [comment]
@manuelaugustin 
> See the screenshot of the code coverage of findKeywordInPageTitle.js as I see it in the report (the implicit else condition isn't covered).

Looks like my coverage test was not updating properly, so I could not see the decrease. Now it is fixed and I added a spec to cover the implicit-else.

> Github seems to be confused again: recent changes in develop (Flesch reading ease, PR about the keywords consisting only of special characters) are also in this PR. If all the branches have been pulled/merged/pushed correctly, probably just changing the base branch to develop and changing it back again to feature/recalibration should fix the issue.

Fixed.

> There are some merge conflicts.

Fixed

> I changed the fixes in the issue description; you need to add fixes before every issue number for Github to recognize it (see https://help.github.com/articles/closing-issues-using-keywords/).

Thanks!

File: src/researches/findKeywordInPageTitle.js [hunk 7]
```diff
@@ -1,24 +1,42 @@
...
+
+	let result = { matches: 0, position: -1 };
 	result.matches = wordMatch( title, keyword, locale );
-	result.position = title.toLocaleLowerCase().indexOf( keyword );
+
+	title = title.toLocaleLowerCase();
+	keyword = keyword.toLocaleLowerCase();
+	result.position = title.indexOf( keyword );
+
+	if ( result.matches > 0 && result.position > 0 ) {
```
Comment (jcomack) [inline src/researches/findKeywordInPageTitle.js]: I'd consider rewriting this so it's an early return:

```
if  ( result.matches === 0 || result.position <= 0 ) {
    return;
}
```

File: src/researches/findKeywordInPageTitle.js [hunk 8]
```diff
@@ -1,24 +1,42 @@
...
  */
 
 module.exports = function( paper ) {
-	var title = paper.getTitle();
-	var keyword = escapeRegExp( paper.getKeyword() );
-	var locale = paper.getLocale();
-	var result = { matches: 0, position: -1 };
+	let title = removePunctuation( paper.getTitle() );
+	let keyword = escapeRegExp( paper.getKeyword() );
+	const locale = paper.getLocale();
```
Comment (jcomack) [inline src/researches/findKeywordInPageTitle.js]: Seeing as this variable is only used once on line 26, I'd suggest removing this line and just call `paper.getLocale()` on line 26 as the third parameter.

File: src/researches/findKeywordInPageTitle.js [hunk 9]
```diff
@@ -1,24 +1,42 @@
...
+	let result = { matches: 0, position: -1 };
 	result.matches = wordMatch( title, keyword, locale );
-	result.position = title.toLocaleLowerCase().indexOf( keyword );
+
+	title = title.toLocaleLowerCase();
+	keyword = keyword.toLocaleLowerCase();
+	result.position = title.indexOf( keyword );
+
+	if ( result.matches > 0 && result.position > 0 ) {
+		const articles = functionWords[ getLanguage( locale ) ].articles;
```
Comment (jcomack) [inline src/researches/findKeywordInPageTitle.js]: I'd suggest adding more documentation as to _what_ is happening here. Without prior knowledge, it's unclear that you're trying to deal with cases where a function word might exist before the actual keyword.

File: src/researches/findKeywordInPageTitle.js (hunk 8 above)
Comment (nataliashitova) [reply src/researches/findKeywordInPageTitle.js]: also line 43

**Comment by Dieterrr** [comment]
Acceptance done 🥇

**Pull request merged**